crates/prover/src/runner.rs

```diff
@@ -202,85 +299,139 @@ where
             .get_prover_last_scanned_l1_height()
             .unwrap_or_else(|_| panic!("Failed to get last scanned l1 height from the ledger db"));
 
-        let mut l1_height = match last_scanned_l1_height {
-            Some(height) => height.0 + 1,
+        let start_l1_height = match last_scanned_l1_height {
+            Some(height) => height.0,
             None => get_initial_slot_height::<Da::Spec>(&self.sequencer_client).await,
         };
 
-        let mut l2_height = self.start_height;
-
         let prover_config = self.prover_config.clone().unwrap();
 
-        let pg_client = match prover_config.db_config {
+        let pg_client = match prover_config.clone().db_config {
             Some(db_config) => {
                 info!("Connecting to postgres");
                 Some(PostgresConnector::new(db_config.clone()).await)
             }
             None => None,
         };
 
-        loop {
-            let da_service = &self.da_service;
-
-            let exponential_backoff = ExponentialBackoffBuilder::new()
-                .with_initial_interval(Duration::from_secs(1))
-                .with_max_elapsed_time(Some(Duration::from_secs(5 * 60)))
-                .build();
-            let last_finalized_height = retry_backoff(exponential_backoff.clone(), || async {
-                da_service
-                    .get_last_finalized_block_header()
-                    .await
-                    .map_err(backoff::Error::transient)
-            })
-            .await?
-            .height();
+        // Save the genesis state root to ledger db
+        if self.start_l2_height == 1 {
+            self.ledger_db
+                .set_l2_state_root(0, &self.state_root.clone())?;
+        }
 
-            if l1_height > last_finalized_height {
-                sleep(Duration::from_secs(1)).await;
-                continue;
+        let (l1_tx, mut l1_rx) = mpsc::channel(1);
+        let da_service = self.da_service.clone();
+        let l1_block_cache = self.l1_block_cache.clone();
+        let l1_handle = tokio::spawn(async move {
+            l1_sync(start_l1_height, da_service, l1_tx, l1_block_cache).await;
+        });
+        tokio::pin!(l1_handle);
+
+        let (l2_tx, mut l2_rx) = mpsc::channel(1);
+        let start_l2_height = self.start_l2_height;
+        let sequencer_client = self.sequencer_client.clone();
+        let sync_blocks_count = self.sync_blocks_count;
+        let l2_handle = tokio::spawn(async move {
```

Comment:
This and the L1 handler are not needed. Just get the future from `l1_sync` and `l2_sync` without awaiting them and use those in the loop/select.

Reply:
erce would explain better but from my understanding, 

if an arm of the select! macro starts execution, await in those arms does not yield execution to other async tasks.

so the arm has to finish execution before execution is yielded to another task. which was not really intended behaviour.

@ercecan

Reply:
No that's actually not true.

A `Future` has both `Ready` and `Pending` state and a `poll` method. `select` calls `poll` on each of those futures and if the future returns `Ready` then it executes the body until the first `.await` which yields control over to the runtime. If the `Future` returns `Pending` it moves on to the arm of the select and the same thing happens.

Reply:
No, that's true. We discovered it with Erce, once `select!` selected a branch it stops polling futures.

Reply:
Yes, `select!` macro will not yield until the branch finishes the execution.
From Tokio Documentation:
> The select! macro runs all branches concurrently on the same task. Because all branches of the select! macro are executed on the same task, they will never run simultaneously. The select! macro multiplexes asynchronous operations on a single task.

https://tokio.rs/tokio/tutorial/select

Reply:
Thanks for clarifying... i didn't know this about select but this is weird IMO